22 list product component design

AbobakrMahdii opened the pull request:
# Title: Feat ListingItems component

## Overview:
This pull request adds a new sell item form component.

## Resolved issue:
- #22

## Description:
This component allows users to upload images and provide information about the items they want to sell. The component saves the form data to the `listing Items` collection in the Firebase Firestore database.

## Notes to reviewer:
- The form currently supports up to four images, but this can be easily extended to support more.
- The form validation is not implemented yet, and will be added in a separate pull request.
- This component had been translated to arabic using nexti8.
- There is a new library had been installed `react-toastify`

## Checklist:
- [x] The code is tested and works as expected.
- [x] The code follows the project's coding conventions.
- [x] The code is well-documented.
- [x] The component is responsive and accessible.
- [x] The translation was tested locally and worked perfect 
- [x] The form submission now is submit the texts not the images until add the user loging and signing in
- [x] The toastifiers is working as expected
- [x] The names and passwords validation is working as expected

## Reviewer Checklist:
- [x] Review the code changes and ensure that they are correct and complete.
- [x] Run the code and ensure that it works as expected.
- [x] Check the code for readability, maintainability, and scalability.
- [x] Verify that the code follows the project's coding conventions.
- [x] Ensure that the component is responsive and accessible.
- [x] Approve or request changes to the pull request.
- [x] make shur the tramsulating is working perfectly.
- [x] make sure the validation is working correctly 
- [x] make sure the notefiers is working as expected

## Screenshots/Recordings:
- English version:
![image](https://user-images.githubusercontent.com/93738112/234410625-2be1fc7e-f645-4c1b-8efc-a9a3c8b08c89.png)

- Arabic version:
![image](https://user-images.githubusercontent.com/93738112/234410804-e4a54d9c-de97-4ffb-a7f4-b004a5de0563.png)

nabilashbat committed:
feat(product-listing): add product-listing file page

nabilashbat committed:
style(product-listing): add a starting helper template

nabilashbat committed:
feat(tailwind.config.js): add custom screen szies

nabilashbat committed:
style(product-list): wrap component with section-wrapper

nabilashbat committed:
style(product-list): minor styling changes

nabilashbat committed:
refacteor: update firebase settings

nabilashbat committed:
feat(signup): add error handling & improve/enhance code

nabilashbat committed:
feat(signIn): setup sign in firebase functionality

nabilashbat committed:
feat(signIn): signed in successfully

nabilashbat committed:
docs(signIn): document the sign in code snippet

nabilashbat committed:
feat(firebase): add google authentication provider

nabilashbat committed:
docs(signup): document the signup module

nabilashbat committed:
refactor(firebase): restruct and refactor firebase files.

nabilashbat committed:
feat: the start of next-auth integration

nabilashbat committed:
feat: add the serviceAccountKey.json file for the firebase-admin functionalities

nabilashbat committed:
fix(firebase): comment the next-auth usage and pause it then update firebase authentication

nabilashbat committed:
add serviceAccountKey.json file to .gitignore

nabilashbat committed:
add serviceAccountKey.json file to .gitignore

nabilashbat committed:
delete serviceAccountKey.json file

AbobakrMahdii committed:
fix: merge remote develop conflicts

AbobakrMahdii committed:
fix: merge remote conflicts

AbobakrMahdii committed:
fix: signin Links import

AbobakrMahdii committed:
feat(sell items): create a new component called sell items

create a new component called sell items so that it will let the user add a new product or service
that he can give it in the website

BREAKING CHANGE: new component created

#22

AbobakrMahdii committed:
test(signin): remove the signin test

remove the signin test so that it effect the commits work becase of  the error insted until fix the
error then recap it

BREAKING CHANGE: test removable

22

AbobakrMahdii committed:
chore(sellitems): add a new image handler

add the default image that will be shown as default before the user add the product images

AbobakrMahdii committed:
refactor: add the highlighter that wi
ll be the header of the sellitems page

AbobakrMahdii committed:
refactor: add labels that will handle images input in sellItems page

AbobakrMahdii committed:
refactor: add the list type selector

AbobakrMahdii committed:
refactor: add the category selector

AbobakrMahdii committed:
refactor: add the rest of the inputs

AbobakrMahdii committed:
refactor: add the form buttons

AbobakrMahdii committed:
refactor: remove the image backgroubd

AbobakrMahdii committed:
refactor: design the images as expected

AbobakrMahdii committed:
design: add the large screens design

AbobakrMahdii committed:
design: fix the large screens design

AbobakrMahdii committed:
design: another large screens design fixes

AbobakrMahdii committed:
design: finish design the rest of the screens requested design

AbobakrMahdii committed:
fix(button): add the uploadImage button style

add the uploadImage button style so that it wasn't there

BREAKING CHANGE: new design in the button component

#22

AbobakrMahdii committed:
refactor(button): update the stories code

update the stories book code so that it meets the button code with the new things

#22

AbobakrMahdii committed:
design: add the upload image design to its button

AbobakrMahdii committed:
design: change the justify between in the location and price span to be gap so that it is shows correctly

AbobakrMahdii committed:
design: change the default image to be a clean image

AbobakrMahdii committed:
design: change the buttons width in the button component to be a full width

AbobakrMahdii committed:
refactor: remove the images input so that it will be only displaying the uploaded image not an input with some style issues fixes

AbobakrMahdii committed:
refactor: change the images src so that it will be handeld from the formData

AbobakrMahdii committed:
refactor: update the form data so that it will handle the hole user list things

AbobakrMahdii committed:
refactor:(button) add the onChange event to the button

AbobakrMahdii committed:
refactor: add the inputHandler which will handle the user inputs

AbobakrMahdii committed:
refactor: add the image upload handler

AbobakrMahdii committed:
refactor: add the form submit handler

AbobakrMahdii committed:
refactor: add the submit handler to the section

AbobakrMahdii committed:
refactor: add tybe and name for all elements in the section

AbobakrMahdii committed:
refactor: add the values to the necessary elements in the section

AbobakrMahdii committed:
refactor: add the onChange event to the needed elements in the section

AbobakrMahdii committed:
refactor: change the section to be form

AbobakrMahdii committed:
refactor: Update the images props in the form with the images sorce to be handeld from the upload image handler and set everyPhoto with its prop in the form

AbobakrMahdii committed:
refactor: change the images spans to be labels so that they will be an inputs

AbobakrMahdii committed:
refactor: add inputs to the images labels to be an inputs so that the user can add images

AbobakrMahdii committed:
refactor: add the correct sorce to the images

AbobakrMahdii committed:
refactor: change the sellitems name to be listing product

AbobakrMahdii committed:
refactor: change the sellitems name to be listing product

AbobakrMahdii committed:
refactor:(dynamic input) add the prop type 'file' to the onOf proptypes to fix the consol error

AbobakrMahdii committed:
refactor:(dynamic input) remove the is requierd  propType to fix the consol error

AbobakrMahdii committed:
refactor: add the firebase database to the code so that the listed item handeld in the firebase

AbobakrMahdii committed:
Merge branch 'develop' of github.com:202212-GIZ-YE-FEW/recoded_student-store_codeForce-cap into 22-list-product-component-design

AbobakrMahdii committed:
feat(listingItems.json): add the locales traanslation files

AbobakrMahdii committed:
refactor(listingPage): add the serversideTranslation to listingPage

AbobakrMahdii committed:
refactor(listingitems): add the first translating to test its working

AbobakrMahdii committed:
docs:(listingitems): add the all english translation needed in the page

AbobakrMahdii committed:
docs:(listingitems): add the rest arabic  translation needed in the page

AbobakrMahdii committed:
docs:(listingitems): active the rest translation to the component

AbobakrMahdii committed:
docs:(listingitems): add missing value to the options in the selectors

AbobakrMahdii committed:
refactor(listing items): run the needed terminals

run the lint fixer, prettier, and test to update all the test and make it secces

#22

AbobakrMahdii committed:
fix(product listing): remove the product listing page

remove the product listing page so that its not needed any more because we change it to be
listingItems page

BREAKING CHANGE: page removable

#22

AbobakrMahdii committed:
fix(listing items): add the hight image controller

add a maximum hight to the images so that when the user upload an image its hight not being larger
than the design

#22

AbobakrMahdii committed:
docs(terminal): runing the necessary termals

AbobakrMahdii committed:
test(listingitems): test added

add the listing item test

#22

AbobakrMahdii committed:
test(listing items): remove the test to another time

nabilashbat commented on src/components/ListingItems/ListingItems.jsx:
```diff
@@ -0,0 +1,266 @@
+    event.preventDefault()
+    try {
+      // Save the form data to the sellItems collection
+      const docRef = await addDoc(collection(db, "listing Items"), formData)
```
It would be great if you validate the inputs for any wrong or even malicious inputs. For example, you can review how I did the validation in the sign-up component 😉

AbobakrMahdii replied:
i didn't understand this point we have to discuss the hole firebase things today in sha'a allah so it's important for us to finish the rest of this web page in this week 😊

nabilashbat commented on src/components/ListingItems/ListingItems.jsx:
```diff
@@ -0,0 +1,266 @@
+  }
+
+  // Image upload handler
+  const uploadedImageHandler = (event, imageField) => {
```
Based on your answer to my question in this and the conversation we had in Discord, I tried to look for an example of how to implement a custom hook to upload images to the Firestore database, give it a try:

```JavaScript

import { useState } from "react"
import { addDoc } from "firebase/firestore"

const useUploadImage = (imageFile, collectionRef) => {
  const [imageUrl, setImageUrl] = useState("")

  const uploadImage = async () => {
    const reader = new FileReader()
    reader.readAsDataURL(imageFile)
    reader.onloadend = async () => {
      const imageData = reader.result
      const docRef = await addDoc(collectionRef, {
        imageUrl: imageData,
      })
      setImageUrl(docRef.id)
    }
  }

  return {
    imageUrl,
    uploadImage,
  }
}
```

To use this custom hook, you would simply pass in the image file and the collection reference to the `useUploadImage` hook. For example:

```JavaScript

const { imageUrl, uploadImage } = useUploadImage(imageFile, collectionRef)

// ...

uploadImage()
```

This would upload the image file to the Firestore database and return the image URL. You can then use the image URL to display the image in your application.

AbobakrMahdii replied:
i tried this way which is not working with the firebase because of that the firestore never collect a files when i tried this way it gives me an error which say you are trying to collect a file to the database this way is wrong so that the fles can only had been uploaded to the storage only

nabilashbat commented on src/components/PhotosGallery/PhotosGallery.jsx:
```diff
@@ -27,7 +27,7 @@ function dropImage(
 function PhotosGallery(props) {
   const [file, setFile] = useState(null)
-  const GALLERY_SVG = "gallery.svg"
+  const GALLERY_SVG = "/images/gallery.svg"
```
I just wanted to thank you here for trying to organize and structure things very well, Thumbs up brother 
👍

AbobakrMahdii replied:
thamk you so much bro

nabilashbat commented on src/components/Signin/Signin.jsx:
```diff
@@ -1,12 +1,70 @@
 import Input from "../input"
+
+// Define validation schema using Yup
+const validationSchema = Yup.object().shape({
```
I came across a highly intriguing idea/methodology employed by the iDev team that adheres to the principle of separation of concerns. They have segregated the validation schema of each file section into separate files within the code. I highly recommend that we adopt this approach for our future work. Therefore, I am leaving this review solely to highlight this point.

Here is the link that I am referring to:
https://github.com/202212-GIZ-YE-FEW/iDev/tree/develop/src/utils

AbobakrMahdii replied:
I think we should have a zoom call there is a lot of missing things for me in the firebase 😂😅

nabilashbat commented on src/components/button/Button.jsx:
```diff
@@ -14,7 +14,9 @@ const BUTTON_STYLES = {
     "h-10 w-40 m-1 p-1 rounded-lg border-2 bg-purple-dark font-light text-[15px] text-white shadow-sm",
   listItem:
-    "h-12 w-72 md:h-14 md:w-80 lg:h-14 lg:w-96 m-1 p-1 rounded-lg border-2 bg-pumpkin font-md text-[17px] text-white shadow-sm",
+    "h-12 w-full md:h-14 lg:h-14 m-1 p-1 rounded-lg border-2 bg-pumpkin font-md text-[17px] text-white shadow-sm",
```
If you find that there are many classes that are identical, what I would recommend is that instead of repeating them, just extract them to a variable and then use that variable instead.

AbobakrMahdii replied:
in this one specifically i tried to do what you said here but the problem is that there is a very small deference between every button so when u try to merge any repeated classes in the hole buttons you will never find that because for example the rounded and the shadow if you tried to merge thim in one variable you wil change one button style so that this button is not having the same style like the rest 😂

nabilashbat commented on src/components/button/Button.stories.jsx:
Even putting the usage of stories in mind, I must say that it's truly amazing, my friend. Keep up the great work!

AbobakrMahdii replied:
Thank you ❤️

nabilashbat commented on src/pages/_app.jsx:
Just return this file to its previous state, since we're no longer using next-auth.

```javascript
import { appWithTranslation } from "next-i18next"

import "../styles/globals.css"

import { appWithTranslation } from "next-i18next"

import "../styles/globals.css"

function MyApp({ Component, pageProps }) {
  return (
    <div className='bg-iceblue'>
      <Component {...pageProps} />
    </div>
  )
}
```

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/listing-page.jsx:
I would recommend that you delete the Word page, since the file itself is inside the pages directory which is very obvious that the listing is a page, So no need to mention the Word page.

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/signup/index.jsx:
```diff
@@ -1,8 +1,8 @@
 import Image from "next/image"
 import Link from "next/link"
 import { useRouter } from "next/router"
+import { useSession } from "next-auth/react"
```
This line would be removed.

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/signup/index.jsx:
```diff
@@ -49,17 +50,22 @@ const limiter = rateLimit({
 */
 
 function Signup() {
+  const { data: session } = useSession()
```
and this line should be removed also.

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/signup/index.jsx:
```diff
@@ -13,11 +13,12 @@ import Button from "@/components/button"
+import waitForEmailVerification from "@/utils/firebase/waitForEmailVerification"
 
 // Define validation schema using Yup
 const validationSchema = Yup.object().shape({
```
This validation schema is not valid, and it's not very helpful it's in fact vulnerable, We need to review it!

AbobakrMahdii replied:
Ok we will have a zoom call today it's so important to discus this

AbobakrMahdii replied:
In progress !!

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/signup/index.jsx:
```diff
@@ -49,17 +50,22 @@ const limiter = rateLimit({
 */
 
```
There should be something to secure and clean the inputs that I've got to add, which is something like:

```
// This function handles changes to the form input fields.
function handleChange(event) {
  const { name, value } = event.target;
  setFormData((prevState) => ({
    ...prevState,
    [name]: DOMPurify.sanitize(value),
  }));
}
```

AbobakrMahdii replied:
Done !!

nabilashbat commented on src/pages/signup/index.jsx:
When it comes to dealing with forms, we should consider implementing certain measures, such as the following, to enhance security.

* **To validate the form, you can use the `useForm` hook from React Hook Form.** For example, you can use the following code to validate the email field:

```Code snippet

const { isValid, errors } = useForm({
  name: 'email',
  rules: {
    required: true,
    email: true,
  },
});

if (!isValid) {
  // Show the errors to the user.
  return <div>{errors.email}</div>;
}
```

* **To secure the form, you can use HTTPS and by hashing passwords before storing them in the database.** For example, you can use the following code to secure the form:

```Code snippet

const { pathname, query } = useLocation();

const handleSubmit = async (e) => {
  e.preventDefault();

  // Validate the form.
  const { isValid, errors } = useForm();

  if (!isValid) {
    // Show the errors to the user.
    return;
  }

  // Get the form data.
  const formData = useForm();

  // Create a new user.
  const newUser = await createUser(formData);

  // Redirect the user to the home page.
  if (newUser) {
    const redirectTo = pathname === '/signup' ? '/' : pathname;
    return redirect(redirectTo);
  }
};
```

AbobakrMahdii replied:
We will be in the future

AbobakrMahdii committed:
refactor(_app): return the _app page to its default code so that we won't use next auth

AbobakrMahdii committed:
refactor(pages): remove the word page from the default pages

AbobakrMahdii committed:
build(sign up): remove the next auth lines

remove the next auth lines with remove next auth library

BREAKING CHANGE: library uninstalled

#22

AbobakrMahdii commented:
Hey @nabily4e-dev i think we should have a zoom call today to cover the hole things about the firebase so that i'm not that strong with it 😊

AbobakrMahdii committed:
fix(signup): remove the dublicated form

AbobakrMahdii committed:
refactor(listingItems):add the DOMPurify

AbobakrMahdii committed:
refactor(listingitems(:add the disapled value to the default options with remove its value to make it more specified

AbobakrMahdii committed:
refactor(listingitems):change the selected value from the options to be just a default value

AbobakrMahdii committed:
refactor(listingitems):remove the disabled  && default Valuse from the options so that change it with the hidden value to never be selected

AbobakrMahdii committed:
build(main projuct): install react toastefy to handle the errors

AbobakrMahdii committed:
feat(listingItems): add the schema validation of the listingItems page

AbobakrMahdii committed:
refactor(listingItems):impelement the schema validation to the listingItems component

AbobakrMahdii committed:
refactor(listingItems):impelement the schema validation to be displayed by the Toastify so that it look good

AbobakrMahdii committed:
refactor(listingItems):impelement the rest schema validation to be displayed by the Toastify

AbobakrMahdii committed:
feat(utils): add the schema validation of the signup page with extract its validation to be in its file in the utils

AbobakrMahdii committed:
refactor(signup): change the errors to be displayed by the toast notification

AbobakrMahdii committed:
refactor(signup): change the errors to be displayed by the toast notification

AbobakrMahdii committed:
Merge branch 'develop' of github.com:202212-GIZ-YE-FEW/recoded_student-store_codeForce-cap into 22-list-product-component-design

AbobakrMahdii committed:
docs(utils): add the listingItems product name validation

in this pull request we added the all necessary validation to the sign in and the listing items and
the rest of the needed component to be validate

BREAKING CHANGE: new library with new utils with extractings and addings

#22

AbobakrMahdii committed:
refactor(signin && out): add the toast with extract the validation

add the toast notification to handle the errors and messages with extract the validation to be in
its file in the utils

BREAKING CHANGE: extracted code

#22

AbobakrMahdii committed:
docs(prettier): run the prettier to fix the test error

AbobakrMahdii commented:
Every thing required in this pull request is now perfect please make sure it's working as expected and let me know if there is any another editing in the codes

AbobakrMahdii committed:
refactor(listingitems) update the listingitems valedation

AbobakrMahdii committed:
refactor(pricefilter) update the price filter to display the errors in the toastify

AbobakrMahdii committed:
refactor(listing items) update the submit handler to display a wait notefycation so that the user know he have to wait

AbobakrMahdii committed:
refactor(listing items) the code so that remove the labels and make the errors handeld by the elements it self by onError events

AbobakrMahdii committed:
refactor(listing items) update the listingitems to handle the images inputs

AbobakrMahdii committed:
fix(aboutus): change the aboutus name to be Aboutus so that fix the building error of the deploy

nabilashbat commented on src/components/Signin/Signin.jsx:
```diff
@@ -1,14 +1,65 @@
 import styles from "./Signin.module.css"
 
+import signIn from "@/utils/firebase/signin"
+import { signinValidation } from "@/utils/schemaValidations/signin"
```
Loved this one, thanks 🙏🏽

AbobakrMahdii replied:
Thank you

nabilashbat commented on src/utils/firebase/firestore/db.js:
Why not deleting this file?
🤔

AbobakrMahdii replied:
it had been deleted bro

nabilashbat reviewed (approved):
Honestly, I am speechless. Just keep up the phenomenal job 👏🏽👏🏽👏🏽
